-[WKWebsiteDataStore _fetchDataRecordsOfTypes:withOptions:completionHandler:] returns wrong data size

szewai opened the pull request:
#### 74ce535b465cc554d0b21ca7ea50d8173b89fc85
<pre>
-[WKWebsiteDataStore _fetchDataRecordsOfTypes:withOptions:completionHandler:] returns wrong data size
<a href="https://bugs.webkit.org/show_bug.cgi?id=248459">https://bugs.webkit.org/show_bug.cgi?id=248459</a>
rdar://problem/102752281

Reviewed by Chris Dumez.

Make sure NetworkStorageManager return size in fetched data records when necessary.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::directorySize):
* Source/WTF/wtf/FileSystem.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::hasLocalStorage):
(WebKit::NetworkProcess::fetchWebsiteData):
(WebKit::NetworkProcess::registrableDomainsWithWebsiteData):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::updateOriginData):
(WebKit::NetworkStorageManager::fetchDataFromDisk):
(WebKit::NetworkStorageManager::fetchData):
(WebKit::NetworkStorageManager::deleteDataOnDisk):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::fetchDataTypesInList):
(WebKit::OriginStorageManager::StorageBucket::fetchDataTypesInListFromDisk):
(WebKit::OriginStorageManager::fetchDataTypesInList):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/257661@main">https://commits.webkit.org/257661@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c79ae52d735b3a4ddb880281aa43d323a015f88

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99562 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8737 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32655 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108924 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169161 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103555 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9322 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86042 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92036 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106841 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105321 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7050 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34011 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/partitioned-service-worker-matchAll.tentative.https.html (failure)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21922 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90211 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2578 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23437 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86099 "Built successfully") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2510 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28927 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8658 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42912 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88970 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5273 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4370 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19897 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
```diff
@@ -374,7 +374,7 @@
 		0F6183311C45BF070072450B /* AirLowerMacros.h in Headers */ = {isa = PBXBuildFile; fileRef = 0F6183281C45BF070072450B /* AirLowerMacros.h */; };
 		0F61833C1C45F62A0072450B /* Foundation.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 51F0EB6105C86C6B00E6DF1B /* Foundation.framework */; };
-		0F61833D1C45F62A0072450B /* JavaScriptCore.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 932F5BD90822A1C700736975 /* JavaScriptCore.framework */; };
+		0F61833D1C45F62A0072450B /* .framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 932F5BD90822A1C700736975 /* .framework */; };
```
Not sure why there are changes on this file, will remove..

cdumez commented on Source/WTF/wtf/FileSystem.cpp:
```diff
@@ -665,6 +665,30 @@ std::optional<uint64_t> fileSize(const String& path)
+        return std::nullopt;
+
+    CheckedUint64 size = 0;
+    for (auto& entry : std::filesystem::recursive_directory_iterator(stdPath)) {
```
You need to pass an `ec` to the iterator constructor. Otherwise, this may throw exceptions and will result in crashes.

cdumez commented on Source/WTF/wtf/FileSystem.cpp:
```diff
@@ -665,6 +665,30 @@ std::optional<uint64_t> fileSize(const String& path)
+    CheckedUint64 size = 0;
+    for (auto& entry : std::filesystem::recursive_directory_iterator(stdPath)) {
+        auto filePath = fromStdFileSystemPath(entry.path());
+        if (entry.is_regular_file())
```
Ditto, you're failing to pass `ec` here so it may throw exceptions.

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -596,34 +596,50 @@ HashSet<WebCore::ClientOrigin> NetworkStorageManager::getAllOrigins()
+    }).iterator->value; 
+
+    for (auto [type, size] : newTypeSizeMap) {
+        auto& currentSize = typeSizeMap.ensure(type, [&]() {
```
No need to call ensure() for a builtin type since we're not trying to avoid construction cost when the key already exists. This ends up being more expensive than the simple:
`typeSizeMap.add(type, 0)`

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -596,34 +596,50 @@ HashSet<WebCore::ClientOrigin> NetworkStorageManager::getAllOrigins()
-Vector<WebsiteData::Entry> NetworkStorageManager::fetchDataFromDisk(OptionSet<WebsiteDataType> targetTypes)
+static void updateOriginData(HashMap<WebCore::SecurityOriginData, OriginStorageManager::DataTypeSizeMap>& originTypes, const WebCore::SecurityOriginData& origin, const OriginStorageManager::DataTypeSizeMap& newTypeSizeMap)
+{
+    auto& typeSizeMap = originTypes.ensure(origin, [&]() {
```
This should just call add(), not ensure. The default constructor for a HashMap is really cheap since it is just a wrapper around a null pointer.

cdumez commented on Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
```diff
@@ -228,10 +228,12 @@ bool OriginStorageManager::StorageBucket::isEmpty()
-    result.add(fetchDataTypesInListFromDisk(types));
+    auto result = fetchDataTypesInListFromDisk(types, shouldComputeSize);
+    auto memoryResult = fetchDataTypesInListFromMemory(types);
+    for (auto it = memoryResult.begin(); it != memoryResult.end(); ++it)
```
Can we just a modern range loop?

something like `for (auto type : memoryResult)`

cdumez reviewed (approved):
LGTM with fixes.

szewai committed:
-[WKWebsiteDataStore _fetchDataRecordsOfTypes:withOptions:completionHandler:] returns wrong data size
https://bugs.webkit.org/show_bug.cgi?id=248459
rdar://problem/102752281

Reviewed by Chris Dumez.

Make sure NetworkStorageManager return size in fetched data records when necessary.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::directorySize):
* Source/WTF/wtf/FileSystem.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::hasLocalStorage):
(WebKit::NetworkProcess::fetchWebsiteData):
(WebKit::NetworkProcess::registrableDomainsWithWebsiteData):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::updateOriginData):
(WebKit::NetworkStorageManager::fetchDataFromDisk):
(WebKit::NetworkStorageManager::fetchData):
(WebKit::NetworkStorageManager::deleteDataOnDisk):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::fetchDataTypesInList):
(WebKit::OriginStorageManager::StorageBucket::fetchDataTypesInListFromDisk):
(WebKit::OriginStorageManager::fetchDataTypesInList):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/257661@main
